migrations/back-1572/back-1572.go

```diff
@@ -0,0 +1,212 @@
+package main
+
+import (
+	"time"
+
+	"github.com/globalsign/mgo"
+	"github.com/globalsign/mgo/bson"
+	"github.com/urfave/cli"
+
+	"github.com/tidepool-org/platform/application"
+	"github.com/tidepool-org/platform/errors"
+	migrationMongo "github.com/tidepool-org/platform/migration/mongo"
+	storeStructuredMongo "github.com/tidepool-org/platform/store/structured/mongo"
+)
+
+func main() {
+	application.RunAndExit(NewMigration())
+}
+
+type Migration struct {
+	*migrationMongo.Migration
+}
+
+func NewMigration() *Migration {
+	return &Migration{
+		Migration: migrationMongo.NewMigration(),
+	}
+}
+
+func (m *Migration) Initialize(provider application.Provider) error {
+	if err := m.Migration.Initialize(provider); err != nil {
+		return err
+	}
+
+	m.CLI().Usage = "BACK-1572: Deduplicate data sets affected by UPLOAD-323/BACK-1379"
+	m.CLI().Description = "BACK-1572: Deduplicate data sets affected by UPLOAD-323/BACK-1379. Specifically,\n" +
+		"   Find all data where the `deviceId` field starts with 'InsOmn', and check whether the Jellyfish\n" +
+		"   generated `id` field matches the expected hash.\n" +
+		"   If the `id` field does not match:\n" +
+		"     * Update the `id` and `_id` fields to the expected hashes\n" +
+		"     * Search for any duplicate documents, and archive the document with the initially incorrect hash"
+	m.CLI().Authors = []cli.Author{
+		{
+			Name:  "Lennart Goedhart",
+			Email: "lennart@tidepool.org",
+		},
+	}
+
+	m.CLI().Action = func(ctx *cli.Context) error {
+		if !m.ParseContext(ctx) {
+			return nil
+		}
+		return m.execute()
+	}
+
+	return nil
+}
+
+func (m *Migration) execute() error {
+	m.Logger().Debug("Deduplicate data sets affected by UPLOAD-323/BACK-1379")
+
+	m.Logger().Debug("Creating data store")
+
+	mongoConfig := m.NewMongoConfig()
+	mongoConfig.Database = "data"
+	mongoConfig.Timeout = 60 * time.Minute
+	dataStore, err := storeStructuredMongo.NewStore(mongoConfig, m.Logger())
+	if err != nil {
+		return errors.Wrap(err, "unable to create data store")
+	}
+	defer dataStore.Close()
+
+	m.Logger().Debug("Creating data session")
+
+	dataSession := dataStore.NewSession("deviceData")
+	defer dataSession.Close()
+
+	hashUpdatedCount, archivedCount := m.migrateOmnipodDocuments(dataSession)
+
+	m.Logger().Infof("Migrated %d duplicate Omnipod documents", hashUpdatedCount)
+	m.Logger().Infof("Archived %d duplicate Omnipod documents", archivedCount)
+
+	return nil
+}
+
+func (m *Migration) migrateOmnipodDocuments(dataSession *storeStructuredMongo.Session) (int, int) {
+	logger := m.Logger()
+
+	logger.Debug("Finding distinct users")
+
+	var userIDs []string
+	var hashUpdatedCount, archivedCount int
+
+	err := dataSession.C().Find(bson.M{}).Distinct("_userId", &userIDs)
```

Comment:
Does it make sense to narrow down the query by listing only datums created after a given date and are active? If it's fast enough maybe it's ok to leave it as is.

Reply:
The deduplication only works on ID - it doesn't care whether data is inactive or not, so we still need to take them into account (a bug, to be sure, but still how it currently works).

Reply:
If that's the case, should the `"_active": true` condition be removed from the queries below?